libs/token/src/src_20.sw

```diff
@@ -0,0 +1,341 @@
+///
+/// # Number of Storage Accesses
+///
+/// * Reads: `1`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::_symbol;
+/// use std::string::String;
+///
+/// storage {
+///     symbol: StorageMap<AssetId, StorageKey<StorageString>> = StorageMap {},
+/// }
+///
+/// fn foo(asset: AssetId) {
+///     let symbol = _symbol(storage.symbol, asset);
+///     assert(symbol.unwrap_or(String::new()).len() != 0);
+/// }
+/// ```
+#[storage(read)]
+pub fn _symbol(
+    symbol_key: StorageKey<StorageMap<AssetId, StorageKey<StorageString>>>,
+    asset: AssetId,
+) -> Option<String> {
+    match symbol_key.get(asset).try_read() {
+        Option::Some(s) => s.read_slice(),
+        Option::None(s) => Option::None,
+    }
+}
+
+/// Returns the number of decimals the asset uses.
+///
+/// # Additional Information
+///
+/// e.g. 8, means to divide the token amount by 100000000 to get its user representation.
+///
+/// # Arguments
+///
+/// * `decimals_key`: [StorageKey<StorageMap<AssetId, u8>>] - The location in storage which the `StorageMap` that stores the decimals of assets is stored.
+/// * `asset`: [AssetId] - The asset of which to query the decimals.
+///
+/// # Returns
+///
+/// * [Option<u8>] - The decimal precision used by `asset`.
+///
+/// # Number of Storage Accesses
+///
+/// * Reads: `1`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::_decimals;
+///
+/// storage {
+///     decimals: StorageMap<AssetId, u8> = StorageMap {},
+/// }
+///
+/// fn foo(asset: AssedId) {
+///     let decimals = _decimals(storage.decimals, asset);
+///     assert(decimals.unwrap_or(0u8) == 8);
+/// }
+/// ```
+#[storage(read)]
+pub fn _decimals(
+    decimals_key: StorageKey<StorageMap<AssetId, u8>>,
+    asset: AssetId,
+) -> Option<u8> {
+    decimals_key.get(asset).try_read()
+}
+
+/// Unconditionally sets the name of an asset.
+///
+/// # Additional Information
+///
+/// NOTE: This does not check whether the asset id provided is valid or already exists.
+///
+/// # Arguments
+///
+/// * `name_key`: [StorageKey<StorageMap<AssetId, StorageKey<StorageString>>>] - The location in storage which the `StorageMap` that stores the names of assets is stored.
+/// * `asset`: [AssetId] - The asset of which to set the name.
+/// * `name`: [String] - The name of the asset.
+///
+/// # Number of Storage Accesses
+///
+/// * Writes: `2`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::{_set_name, _name};
+/// use std::string::String;
+///
+/// storage {
+///     name: StorageMap<AssetId, StorageKey<StorageString>> = StorageMap {},
+/// }
+///
+/// fn foo(asset: AssetId) {
+///     let name = String::from_ascii_str("Ether");
+///     _set_name(storage.name, asset, name);
+///     assert(_name(storage.name, asset) == name);
+/// }
+/// ```
+#[storage(write)]
+pub fn _set_name(
+    name_key: StorageKey<StorageMap<AssetId, StorageKey<StorageString>>>,
+    asset: AssetId,
+    name: String,
+) {
+    let name_string_key: StorageKey<StorageString> = StorageKey {
```

Comment:
Would you be interested in using your `new()` syntax here?

Reply:
Sure, but that isn't included in the latest release. Still waiting on forc `v0.46.0`